File: packages/react-native-nitro-fetch/ios/FetchCache.swift
import Foundation

final class FetchCache {
  struct CachedEntry {
    let response: NitroResponse
    let timestampMs: Double
  }

  private static let lock = NSLock()
  private static var pending: [String: [(Result<NitroResponse, Error>) -> Void]] = [:]
  private static var results: [String: CachedEntry] = [:]

  static func getPending(_ key: String) -> Bool {
    lock.lock()
    defer { lock.unlock() }
    return pending[key] != nil
  }

  static func beginPending(_ key: String) -> Bool {
    lock.lock()
    defer { lock.unlock() }
    if pending[key] != nil { return false }
    pending[key] = []
    return true
  }


  static func joinPending(
    _ key: String,
    completion: @escaping (Result<NitroResponse, Error>) -> Void
  ) -> Bool {
    lock.lock()
    defer { lock.unlock() }
    guard pending[key] != nil else { return false }
    pending[key]?.append(completion)
    return true
  }

  static func complete(_ key: String, with result: Result<NitroResponse, Error>) {
    lock.lock()
    let callbacks = pending.removeValue(forKey: key) ?? []
    if case let .success(resp) = result {
      results[key] = CachedEntry(response: resp, timestampMs: ProcessInfo.processInfo.systemUptime * 1000)
    }
    lock.unlock()
    // Outside the lock: a callback may re-enter FetchCache.
    callbacks.forEach { $0(result) }
  }

  static func getResultIfFresh(_ key: String, maxAgeMs: Double) -> NitroResponse? {
    lock.lock()
    defer { lock.unlock() }
    guard let entry = results[key] else { return nil }
    let age = ProcessInfo.processInfo.systemUptime * 1000 - entry.timestampMs
    if maxAgeMs.isFinite && maxAgeMs > 0 && age <= maxAgeMs { return entry.response }
    results.removeValue(forKey: key)
    return nil
  }
}

File: packages/react-native-nitro-fetch/android/src/main/java/com/margelo/nitro/nitrofetch/FetchCache.kt
package com.margelo.nitro.nitrofetch

import java.util.concurrent.CompletableFuture
import java.util.concurrent.ConcurrentHashMap

data class CachedEntry(val response: NitroResponse, val timestampMs: Long)

object FetchCache {
  private val pending = ConcurrentHashMap<String, CompletableFuture<NitroResponse>>()
  private val results = ConcurrentHashMap<String, CachedEntry>()

  fun getPending(key: String): CompletableFuture<NitroResponse>? = pending[key]

  /** Atomically registers a prefetch. Returns the already-pending future if one exists. */
  fun beginPending(key: String, future: CompletableFuture<NitroResponse>): CompletableFuture<NitroResponse>? {
    val existing = pending.putIfAbsent(key, future)
    if (existing == null) {
      future.whenComplete { _, _ -> pending.remove(key, future) }
    }
    return existing
  }

  fun complete(key: String, value: NitroResponse) {
    results[key] = CachedEntry(value, System.nanoTime() / 1_000_000)
    pending.remove(key)?.complete(value)
  }

  fun completeExceptionally(key: String, t: Throwable) {
    pending.remove(key)?.completeExceptionally(t)
  }

  fun getResult(key: String): NitroResponse? {
    val entry = results.remove(key) ?: return null
    return entry.response
  }

  fun getResultIfFresh(key: String, maxAgeMs: Long): NitroResponse? {
    val entry = results.remove(key) ?: return null
    val age = System.nanoTime() / 1_000_000 - entry.timestampMs
    return if (maxAgeMs > 0 && age <= maxAgeMs) entry.response else null
  }

  /**
   * Check if a fresh result exists WITHOUT consuming it.
   * Used to check if we should skip starting a new prefetch.
   */
  fun hasFreshResult(key: String, maxAgeMs: Long): Boolean {
    val entry = results[key] ?: return false
    val age = System.nanoTime() / 1_000_000 - entry.timestampMs
    if (maxAgeMs > 0 && age <= maxAgeMs) return true
    results.remove(key, entry)
    return false
  }

  fun clear() {
    pending.clear()
    results.clear()
  }
}
